backends/dpdk/dpdkArch.cpp

```diff
@@ -467,6 +467,215 @@ bool CollectMetadataHeaderInfo::preorder(const IR::Type_Struct *s) {
+                            offset += s->second.fieldWidth;
+                        }
+                        fields->push_back(new IR::StructField(IR::ID(modifiedName),
+                                          IR::Type_Bits::get(size_sum_so_far)));
+                        size_sum_so_far = 0;
+                        modifiedName = "";
+                        field_name_list.clear();
+                    }
+                } else {
+                    ::error("8-bit Alignment for Header Structure '%1%' is not possible",
+                            st->name.name);
+                    return st;
+                }
+            } else {
+                fields->push_back(field);
+            }
+        }
+        /* Throw error if there is non-aligned field present in header */
+        if (size_sum_so_far != 0) {
+            ::error("8-bit Alignment for Header Structure '%1%' is not possible", st->name.name);
+            return st;
+        }
+        return new IR::Type_Header(IR::ID(st->name), st->annotations, *fields);
+    }
+
+    return st;
+}
+
+/* This function process the occurrence of non-aligned fields and replaces it with
+   modified member with proper lsb and msb as a slice Member 
+   
+   For eg : If below header
+       header ipv4_t {
+            ...
+            bit<3>  flags;
+            bit<13> fragOffset;
+            ...
+       }
+       is converted into
+       header ipv4_t {
+            ...
+            bit<16>  flags_fragOffset;
+            ...
+       }
+       and there is a field used in P4 as mentioned below
+           if (hdrs.ipv4.fragOffset == 4) ....
+       
+       Then, the above condition is converted into
+           if (hdrs.ipv4.flags_fragOffset[15:3] == 4) .... 
+*/
+const IR::Node *AlignHdrMetaField::preorder(IR::Member *m) {
+    cstring hdrStrName = "";
+    auto it = structure->hdrFieldInfoList.find(m->member.name);
+    if (it != structure->hdrFieldInfoList.end()) {
+        for (auto memVec : structure->hdrFieldInfoList[m->member.name]) {
+            if ((m != nullptr) && (m->expr != nullptr) &&
+                (m->expr->type != nullptr) &&
+                (m->expr->type->is<IR::Type_StructLike>())) {
+                auto str_type = m->expr->type->to<IR::Type_StructLike>();
+                hdrStrName = str_type->name.name;
+            } else {
+                return m;
+            }
+            if (memVec.headerStr != hdrStrName)
+                continue;
+            auto mem = new IR::Member(m->expr, IR::ID(memVec.modifiedName));
+            auto sliceMem = new IR::Slice(mem->clone(), (memVec.offset + memVec.fieldWidth - 1),
+                                          memVec.offset);
+            return sliceMem;
+        }
+    }
+    return m;
+}
+
+/* This function processes the metadata structure and modify the metadata field width
+   to 32/64 bits if it is not 8-bit aligned */
+const IR::Node* ReplaceHdrMetaField::postorder(IR::Type_Struct *st) {
```

Comment:
Where do you fix all the code that accesses the struct?

Reply:
For metadata, since we are upgrading the bit width to 32/64 bits in the structure and directly emitting it into assembly IR,
there is no TypeChecking required for the usage in assembly IR.